sklearn/preprocessing/discretization.py

```diff
@@ -0,0 +1,217 @@
+        cat_features = column_or_1d(cat_features)
+
+        if len(set(cat_features)) != cat_features.shape[0]:
+            raise ValueError("Duplicate categorical column indexes found.")
+
+        if np.all(cat_features >= 0) and np.all(cat_features < n_features):
+            return
+
+        raise ValueError("Invalid categorical feature index.")
+
+    def _cut_points_from_index(self, col_idx):
+        """Determines the cut points of a feature based on the column index.
+        """
+        n_bins = self.n_bins
+        if col_idx in self.categorical_features_set_:
+            return [np.nan] * (self.n_bins - 1)
+
+        _min = self.mins_[col_idx]
+        _max = self.maxes_[col_idx]
+        spacing = (_max - _min) / n_bins
+        return np.linspace(_min + spacing, _max - spacing, num=n_bins - 1)
+
+    def _fit(self, X):
+        """Fits the attributes of the discretizer.
+        """
+        n_features = X.shape[1]
+        if self.categorical_features:
+            self.categorical_features_set_ = set(self.categorical_features)
+        else:
+            self.categorical_features_set_ = {}
+
+        self.mins_ = np.min(X, axis=0)
+        self.maxes_ = np.max(X, axis=0)
+
+        if np.any(self.mins_ - self.maxes_ == 0):
+            raise ValueError("Array contained features which contained the "
+                             "same min and max values.")
+
+        cut_points = [self._cut_points_from_index(i)
+                      for i in xrange(n_features)]
+
+        self.cut_points_ = np.array(cut_points).T
+        self.n_features_ = n_features
+
+    def transform(self, X, y=None):
+        """Discretizes the data. Leaves any categorical features as is.
+
+        Parameters
+        ----------
+        X : numeric array-like, shape (n_samples, n_features)
+            Data containing features to be discretized.
+
+        Returns
+        -------
+        T : numeric array-like, shape (n_samples, n_features)
+            Categorical features will not be transformed. If all features
+            are continuous, T will be an int array. Otherwise T will be
+            a float array.
+
+        """
+        check_is_fitted(self, ["mins_", "maxes_", "n_features_",
+                               "cut_points_"])
+
+        X = check_array(X, dtype=float, ensure_2d=False)
+
+        if X.ndim == 1:
+            raise ValueError("Input array must be 2D shape.")
+
+        if X.shape[1] != self.n_features_:
+            raise ValueError("Transformed array does not have correct number "
+                             "of features. Expecting {0}, received {1}."
+                             .format(self.n_features_, X.shape[1]))
+
+        return self._transform(X)
+
+    def _transform_from_index(self, X, col_idx):
+        X_col = X[:, col_idx]
+        if col_idx in self.categorical_features_set_:
+            return X_col
+        return np.digitize(X_col, self.cut_points_[:, col_idx])
```

Comment:
As long as we're dealing with fixed-width bins, we should be able to do the discretization without `digitize`; division, subtraction, truncation and clipping should be sufficient `O(n_values)` rather than `O(n_values * n_bins)`.